a)Carga los datos y crea un DataFrame con cada uno de ellos.

importando los datos de los documentos .pkl y obtengo los nombres de las columnas para ver la relacion que tienen obtengo que se relacionan por el numero de caso


In [107]:
import pandas as pd

incidents_df = pd.read_pickle("incidents.pkl")
officers_df = pd.read_pickle("officers.pkl")
subjects_df = pd.read_pickle("subjects.pkl")

#obtengo los nombres de las columnas de cada dataframe
column_names = incidents_df.columns
column_names1 = officers_df.columns
column_names2 = subjects_df.columns
#muestro el primer dataframe incidents
column_names

Index(['case_number', 'date', 'location', 'subject_statuses', 'subject_weapon',
       'subjects', 'subject_count', 'officers', 'officer_count',
       'grand_jury_disposition', 'attorney_general_forms_url', 'summary_url',
       'summary_text', 'latitude', 'longitude'],
      dtype='object')

muestros el segundo data frame officers

In [108]:
column_names1

Index(['case_number', 'race', 'gender', 'last_name', 'first_name',
       'full_name'],
      dtype='object')

muestro el tercer dataframe subject

In [109]:
column_names2

Index(['case_number', 'race', 'gender', 'last_name', 'first_name',
       'full_name'],
      dtype='object')

puedo ver que la relaciones entre las 3 es la columna case_number

b. Genera una tabla que contenga la unión de las 3 tablas. hint: utiliza sufijos para para las
columnas que se llaman igual usando el parámetro suffixes de pd.merge().

hago un merge anidado de las tablas, primero hago un merge entre la tabla insidente y official con relacion en el numero de caso al igual que las tablas incidente y sujeto y despues le hago otro mergue para que las 3 tablas esten unidas por el numero de caso


In [110]:
merged_df = pd.merge(
    pd.merge(
        incidents_df, officers_df, on="case_number", suffixes=("_incident", "_officer")
    ),
    subjects_df,
    on="case_number",
    suffixes=("_incident", "_subject"),
)
merged_df
#muestro el resultado del merge pero solo las titulos de las columnas y veo que al hacer el merge cambio los nombre de las columnas que eran repetidos con los subject y coloco 
#incident en vez de officer
merged_df.columns

Index(['case_number', 'date', 'location', 'subject_statuses', 'subject_weapon',
       'subjects', 'subject_count', 'officers', 'officer_count',
       'grand_jury_disposition', 'attorney_general_forms_url', 'summary_url',
       'summary_text', 'latitude', 'longitude', 'race_incident',
       'gender_incident', 'last_name_incident', 'first_name_incident',
       'full_name_incident', 'race_subject', 'gender_subject',
       'last_name_subject', 'first_name_subject', 'full_name_subject'],
      dtype='object')

renombro los nombres para identificarlos mejor 

In [111]:
merged_df.rename(
    columns={
        "race_incident": "race_officer",
        "gender_incident": "gender_officer",
        "last_name_incident": "last_name_officer",
        "first_name_incident": "first_name_officer",
        "full_name_incident": "full_name_officer",
    },
    inplace=True,
)

como ya se que la relacion esta en la columna casa number entonces seria como el id de cada caso o columna por lo que cada colmna me representa un caso por lo que nose debe repetir para eso borro los casos duplicados por numero de caso


c. Verifica si hay filas duplicadas; si es así, elimínalas

In [112]:
#verifico cuales son las duplicadas para despues borrarla en la otro linea de codigo
duplicated = merged_df.duplicated(subset=["case_number"])
merged_df[duplicated]

case_number        date               location subject_statuses  \
1         44523A  2013-02-23  3000 Chihuahua Street          Injured   
5        384832T  2007-05-26        7900 S. Loop 12   Shoot and Miss   
6        384832T  2007-05-26        7900 S. Loop 12   Shoot and Miss   
9        363184P  2005-05-09   4900 Eastside Avenue   Shoot and Miss   
11       559487M  2003-07-24        7728 S. Loop 12         Deceased   
..           ...         ...                    ...              ...   
372  165193-2016  2016-07-07        801 Main Street         Deceased   
373  165193-2016  2016-07-07        801 Main Street         Deceased   
374  165193-2016  2016-07-07        801 Main Street         Deceased   
375  165193-2016  2016-07-07        801 Main Street         Deceased   
376  165193-2016  2016-07-07        801 Main Street         Deceased   

    subject_weapon                                           subjects  \
1          Handgun                                   Curry, James L/M   
5          Unarmed  Smith, James B/M; Dews, Antonio B/M; Spearman,...   
6          Unarmed  Smith, James B/M; Dews, Antonio B/M; Spearman,...   
9          Vehicle                                  Garcia, David L/M   
11         Handgun                                   Vernon, Tony B/M   
..             ...                                                ...   
372  Assault Rifle                                 Johnson, Micah B/M   
373  Assault Rifle                                 Johnson, Micah B/M   
374  Assault Rifle                                 Johnson, Micah B/M   
375  Assault Rifle                                 Johnson, Micah B/M   
376  Assault Rifle                                 Johnson, Micah B/M   

     subject_count                                           officers  \
1                1          Patino, Michael L/M; Fillingim, Brian W/M   
5                3                                 Mondy, Michael B/M   
6                3                                 Mondy, Michael B/M   
9                1                Woods, Jesse W/M; Culp, Barrick B/M   
11               1         Rangel, Mark  L/M; Armendariz, Michael L/M   
..             ...                                                ...   
372              1  Edwards, Henry W/M; Wells, Giovanni B/M; Junge...   
373              1  Edwards, Henry W/M; Wells, Giovanni B/M; Junge...   
374              1  Edwards, Henry W/M; Wells, Giovanni B/M; Junge...   
375              1  Edwards, Henry W/M; Wells, Giovanni B/M; Junge...   
376              1  Edwards, Henry W/M; Wells, Giovanni B/M; Junge...   

     officer_count grand_jury_disposition  ... race_officer gender_officer  \
1                2                No Bill  ...            W              M   
5                1                   None  ...            B              M   
6                1                   None  ...            B              M   
9                2                   None  ...            B              M   
11               2                No Bill  ...            L              M   
..             ...                    ...  ...          ...            ...   
372             12                   None  ...            W              M   
373             12                   None  ...            W              M   
374             12                   None  ...            W              M   
375             12                   None  ...            W              M   
376             12                   None  ...            B              M   

    last_name_officer  first_name_officer    full_name_officer race_subject  \
1           Fillingim               Brian     Fillingim, Brian            L   
5               Mondy             Michael       Mondy, Michael            B   
6               Mondy             Michael       Mondy, Michael            B   
9                Culp             Barrick        Culp, Barrick            L   
11         Armendariz             Michael  Armendariz,

Borrado de duplicados

In [113]:
merged_df = merged_df.drop_duplicates(subset=["case_number"])
merged_df

case_number        date                  location subject_statuses  \
0         44523A  2013-02-23     3000 Chihuahua Street          Injured   
2        121982X  2010-05-03  1300 N. Munger Boulevard          Injured   
3        605484T  2007-08-12   200 S. Stemmons Freeway            Other   
4        384832T  2007-05-26           7900 S. Loop 12   Shoot and Miss   
7        244659R  2006-04-03        6512 South Loop 12          Injured   
..           ...         ...                       ...              ...   
361  161616-2016  2016-07-03   7440 Chesterfield Drive   Shoot and Miss   
362  141461-2016  2016-06-11         2700 Custer Drive   Shoot and Miss   
363  089985-2016  2016-04-16         4800 Columbia Ave   Shoot and Miss   
364  177645-2016  2016-07-23    1716 Chattanooga Place   Shoot and Miss   
365  165193-2016  2016-07-07           801 Main Street         Deceased   

    subject_weapon                                           subjects  \
0          Handgun                                   Curry, James L/M   
2          Handgun                                Chavez, Gabriel L/M   
3          Shotgun                                  Salinas, Nick L/M   
4          Unarmed  Smith, James B/M; Dews, Antonio B/M; Spearman,...   
7            Hands                                 Watkins, Caleb B/M   
..             ...                                                ...   
361        Handgun                                  Brown, Desroy B/M   
362        Vehicle                                        Unknown B/M   
363        Handgun                                        Unknown L/M   
364        Handgun                                        Unknown B/M   
365  Assault Rifle                                 Johnson, Micah B/M   

     subject_count                                           officers  \
0                1          Patino, Michael L/M; Fillingim, Brian W/M   
2                1                               Padilla, Gilbert L/M   
3                1                                  Poston, Jerry W/M   
4                3                                 Mondy, Michael B/M   
7                1                             Armstrong, Michael W/M   
..             ...                                                ...   
361              1                               Crenshaw, Joshua W/M   
362              1                                 Moore, Douglas B/M   
363              1                               Ruben, Fredirick W/M   
364              1                                   Rosen, Brett W/M   
365              1  Edwards, Henry W/M; Wells, Giovanni B/M; Junge...   

     officer_count grand_jury_disposition  ... race_officer gender_officer  \
0                2                No Bill  ...            L              M   
2                1                No Bill  ...            L              M   
3                1            See Summary  ...            W              M   
4                1                   None  ...            B              M   
7                1                No Bill  ...            W              M   
..             ...                    ...  ...          ...            ...   
361              1                   None  ...            W              M   
362              1                   None  ...            B              M   
363              1                   None  ...            W              M   
364              1                   None  ...            W              M   
365             12                   None  ...            W              M   

    last_name_officer  first_name_officer   full_name_officer race_subject  \
0              Patino             Michael     Patino, Michael            L   
2             Padilla             Gilbert    Padilla, Gilbert            L   
3              Poston               Jerry       Poston, Jerry            L   
4               Mondy             Michael      Mondy, Michael            B   
7           Armstrong  

d. ¿Cuántos sujetos de género F hay en el DataFrame resultante? hint: usa el método
.value_counts() sobre la columna.

In [117]:
#opcion 1 con value_counts
female_subjects_count = merged_df["gender_officer"].value_counts()["F"] + merged_df["gender_subject"].value_counts()["F"]
# opcion 2 sin value_counts
#female_subjects_count2 = (merged_df["gender_officer"] == "F").sum()+(merged_df["gender_subject"] == "F").sum()
# female_subjects_count
female_subjects_count

14

vemos la cuentas de todas Feminas del data frame tanto oficiales como ssopechosas 

e. ¿En cuántos números de caso hay por lo menos una sospechosa que sea mujer? hint:
utiliza el método unique() para obtener los valores únicos de una columna específica de
un DataFrame luego de filtrar.

In [118]:
cases_with_female_suspects = merged_df[merged_df["gender_subject"] == "F"][
    "case_number"
].unique()
count_cases_with_female_suspects = len(cases_with_female_suspects)
count_cases_with_female_suspects

6

f. Genera una tabla pivote que muestre el número de casos por mes en la fila y por año en
las columnas. hint: genere columnas de tipo fecha con pd.to_datetime(). ¿Se ve alguna
relación entre mes del año y cantidad de casos?

In [86]:
merged_df["date"] = pd.to_datetime(merged_df["date"])
colum_names4 = merged_df.columns
colum_names4

Index(['case_number', 'date', 'location', 'subject_statuses', 'subject_weapon',
       'subjects', 'subject_count', 'officers', 'officer_count',
       'grand_jury_disposition', 'attorney_general_forms_url', 'summary_url',
       'summary_text', 'latitude', 'longitude', 'race_incident',
       'gender_incident', 'last_name_incident', 'first_name_incident',
       'full_name_incident', 'race_subject', 'gender_subject',
       'last_name_subject', 'first_name_subject', 'full_name_subject', 'year',
       'month', 'month_name'],
      dtype='object')

In [90]:
merged_df["year"] = merged_df["date"].dt.year
merged_df["month"] = merged_df["date"].dt.strftime("%B")

In [91]:
pivot_table_month_year = pd.pivot_table(
    merged_df,
    index="month",
    columns="year",
    aggfunc="count",
    values="case_number",
    fill_value=0,
)
pivot_table_month_year

year       2003  2004  2005  2006  2007  2008  2009  2010  2011  2012  2013  \
month                                                                         
April         1     4     0     2     0     1     0     1     0     0     3   
August        2     0     1     0     3     4     0     0     0     3     0   
December      0     3     4     0     0     2    15     2     5     3     1   
February      0     1     1     7     1     0     0     3     0     7     7   
January       2     2     0     1     1     1     2     0     0     0     7   
July          4     2     0     0     1     4     0     0     1     2     2   
June          8     1     0     0     3     3     0     2     2     2     3   
March         1     8     1     1     5     8     3     7     2     6     3   
May           1     1     2     2     9     5     4     1     1     9     1   
November      3     0     2     8     0     0     1     8     3     1     1   
October       7     1     0     0     0     0     3     1     1     2     4   
September     0     1     0     3     2     2     4     0     0     0     5   

year       2014  2015  2016  
month                        
April         4     0     1  
August        7     9     1  
December      1     0     0  
February      3     1     0  
January       4     0     0  
July          2     4    14  
June          6    18     2  
March         0     0     1  
May           1     0     0  
November      0     5     0  
October       2     0     0  
September     1     3     0

Genera una tabla pivote que muestre en las filas el género del oficial y en las columnas el
género del subject. ¿Cómo interpretas los valores que muestra esta vista?


In [106]:
gender_pivot_table = pd.pivot_table(
    merged_df,
    index="gender_officer",
    columns="gender_subject",
    aggfunc="count",
    values="case_number",
)
gender_pivot_table

gender_subject  F    M
gender_officer        
F               2   18
M               7  350

en 2 casos la supechosa fue mujer y el oficial fue mujer  
en 18 casos la oficial mujer y el sospechoso fue hombre
en 7 casos el oficial fue hombre y el sospechoso mujer 
en 350 el oficial fue hombre y el sospechoso fue hombre 

                                            PARTE 2

2. Para continuar con el desarrollo de este desafío, necesitarás el archivos
Cleaned_DS_Jobs.csv

a. Carga los datos y crea un DataFrame con ellos

In [120]:
import pandas as pd
import numpy as np

# Cargar el archivo CSV y crear un DataFrame
df_jobs = pd.read_csv('Cleaned_DS_Jobs.csv')
df_jobs

Job Title Salary Estimate  \
0    Sr Data Scientist        137-171    
1       Data Scientist        137-171    
2       Data Scientist        137-171    
3       Data Scientist        137-171    
4       Data Scientist        137-171    
..                 ...             ...   
655     Data Scientist        105-167    
656     Data Scientist        105-167    
657     Data Scientist        105-167    
658     Data Scientist        105-167    
659     Data Scientist        105-167    

                                       Job Description  Rating  \
0    Description\n\nThe Senior Data Scientist is re...     3.1   
1    Secure our Nation, Ignite your Future\n\nJoin ...     4.2   
2    Overview\n\n\nAnalysis Group is one of the lar...     3.8   
3    JOB DESCRIPTION:\n\nDo you have a passion for ...     3.5   
4    Data Scientist\nAffinity Solutions / Marketing...     2.9   
..                                                 ...     ...   
655  Summary\n\nWe’re looking for a data scientist ...     3.6   
656  Job Description\nBecome a thought leader withi...     0.0   
657  Join a thriving company that is changing the w...     0.0   
658  100 Remote Opportunity As an AINLP Data Scient...     5.0   
659  Description\n\nThe Data Scientist will be part...     2.7   

             Company Name           Location            Headquarters  \
0             Healthfirst       New York, NY            New York, NY   
1                 ManTech      Chantilly, VA             Herndon, VA   
2          Analysis Group         Boston, MA              Boston, MA   
3                 INFICON         Newton, MA  Bad Ragaz, Switzerland   
4      Affinity Solutions       New York, NY            New York, NY   
..                    ...                ...                     ...   
655              TRANZACT       Fort Lee, NJ            Fort Lee, NJ   
656                  JKGT  San Francisco, CA                      -1   
657            AccessHope      Irwindale, CA                      -1   
658  ChaTeck Incorporated  San Francisco, CA         Santa Clara, CA   
659         1-800-Flowers       New York, NY         Carle Place, NY   

                        Size        Type of ownership  \
0     1001 to 5000 employees   Nonprofit Organization   
1    5001 to 10000 employees         Company - Public   
2     1001 to 5000 employees  Private Practice / Firm   
3      501 to 1000 employees         Company - Public   
4        51 to 200 employees        Company - Private   
..                       ...                      ...   
655   1001 to 5000 employees        Company - Private   
656                       -1                       -1   
657                       -1                       -1   
658        1 to 50 employees        Company - Private   
659   1001 to 5000 employees         Company - Public   

                                  Industry  ... company_age python  excel  \
0                       Insurance Carriers  ...          27      0      0   
1                   Research & Development  ...          52      0      0   
2                               Consulting  ...          39      1      1   
3    Electrical & Electronic Manufacturing  ...          20      1      1   
4                  Advertising & Marketing  ...          22      1      1   
..                                     ...  ...         ...    ...    ...   
655                Advertising & Marketing  ...          31      1      1   
656                                     -1  ...          -1      0      0   
657                                     -1  ...          -1      1      1   
658                Advertising & Marketing  ...          -1      1      0   
659                              Wholesale  ...          44      1      1   

     hadoop  spark aws  tableau  big_data        job_simp  seniority  
0         0      0   1        0         0  data scientist     senior  
1         1      0   0        0         1  data scientist         na  
2         0      0   1     

b. Utiliza la siguiente lista de valores que serán considerados como nulos:
["na", "NA", -1, "0", "-1", "null", "n/a", "N/A", "NULL"]


In [123]:
# Lista de valores considerados como nulos
null_values = ["na", "NA", -1, "0", "-1", "null", "n/a", "N/A", "NULL"]
null_values

['na', 'NA', -1, '0', '-1', 'null', 'n/a', 'N/A', 'NULL']

(hint: utiliza el método replace para reemplazar los valores indicados por np.nan)

In [124]:
# Reemplazar los valores nulos por NaN
df_jobs.replace(null_values, np.nan, inplace=True)
df_jobs

Job Title Salary Estimate  \
0    Sr Data Scientist        137-171    
1       Data Scientist        137-171    
2       Data Scientist        137-171    
3       Data Scientist        137-171    
4       Data Scientist        137-171    
..                 ...             ...   
655     Data Scientist        105-167    
656     Data Scientist        105-167    
657     Data Scientist        105-167    
658     Data Scientist        105-167    
659     Data Scientist        105-167    

                                       Job Description  Rating  \
0    Description\n\nThe Senior Data Scientist is re...     3.1   
1    Secure our Nation, Ignite your Future\n\nJoin ...     4.2   
2    Overview\n\n\nAnalysis Group is one of the lar...     3.8   
3    JOB DESCRIPTION:\n\nDo you have a passion for ...     3.5   
4    Data Scientist\nAffinity Solutions / Marketing...     2.9   
..                                                 ...     ...   
655  Summary\n\nWe’re looking for a data scientist ...     3.6   
656  Job Description\nBecome a thought leader withi...     0.0   
657  Join a thriving company that is changing the w...     0.0   
658  100 Remote Opportunity As an AINLP Data Scient...     5.0   
659  Description\n\nThe Data Scientist will be part...     2.7   

             Company Name           Location            Headquarters  \
0             Healthfirst       New York, NY            New York, NY   
1                 ManTech      Chantilly, VA             Herndon, VA   
2          Analysis Group         Boston, MA              Boston, MA   
3                 INFICON         Newton, MA  Bad Ragaz, Switzerland   
4      Affinity Solutions       New York, NY            New York, NY   
..                    ...                ...                     ...   
655              TRANZACT       Fort Lee, NJ            Fort Lee, NJ   
656                  JKGT  San Francisco, CA                     NaN   
657            AccessHope      Irwindale, CA                     NaN   
658  ChaTeck Incorporated  San Francisco, CA         Santa Clara, CA   
659         1-800-Flowers       New York, NY         Carle Place, NY   

                        Size        Type of ownership  \
0     1001 to 5000 employees   Nonprofit Organization   
1    5001 to 10000 employees         Company - Public   
2     1001 to 5000 employees  Private Practice / Firm   
3      501 to 1000 employees         Company - Public   
4        51 to 200 employees        Company - Private   
..                       ...                      ...   
655   1001 to 5000 employees        Company - Private   
656                      NaN                      NaN   
657                      NaN                      NaN   
658        1 to 50 employees        Company - Private   
659   1001 to 5000 employees         Company - Public   

                                  Industry  ... company_age python  excel  \
0                       Insurance Carriers  ...        27.0      0      0   
1                   Research & Development  ...        52.0      0      0   
2                               Consulting  ...        39.0      1      1   
3    Electrical & Electronic Manufacturing  ...        20.0      1      1   
4                  Advertising & Marketing  ...        22.0      1      1   
..                                     ...  ...         ...    ...    ...   
655                Advertising & Marketing  ...        31.0      1      1   
656                                    NaN  ...         NaN      0      0   
657                                    NaN  ...         NaN      1      1   
658                Advertising & Marketing  ...         NaN      1      0   
659                              Wholesale  ...        44.0      1      1   

     hadoop  spark aws  tableau  big_data        job_simp  seniority  
0         0      0   1        0         0  data scientist     senior  
1         1      0   0        0         1  data scientist        NaN  
2         0      0   1     

c. Elimina todas las filas con datos faltantes. (hint: utiliza el método .dropna())


In [127]:
# Eliminar filas con datos faltantes
df_jobs.dropna(inplace=True)
df_jobs

Job Title Salary Estimate  \
0                               Sr Data Scientist        137-171    
32   Senior Research Statistician- Data Scientist         75-131    
38                  Senior Analyst/Data Scientist         75-131    
45                          Senior Data Scientist         75-131    
54                          Senior Data Scientist         75-131    
..                                            ...             ...   
579                          Senior Data Engineer        138-158    
583                         Senior Data Scientist         80-132    
601                         Senior Data Scientist         87-141    
640                         Senior Data Scientist        105-167    
641        Sr. ML/Data Scientist - AI/NLP/Chatbot        105-167    

                                       Job Description  Rating  \
0    Description\n\nThe Senior Data Scientist is re...     3.1   
32   Acuity is seeking a Senior Research Statistici...     4.8   
38   At Edmunds were driven to make car buying easi...     3.4   
45   Klaviyo is looking for Senior Data Scientists ...     4.8   
54   Benson Hill empowers innovators to develop mor...     3.5   
..                                                 ...     ...   
579  Senior Data Engineer\n\nMaster’s degree in Inf...     2.9   
583  Job Requisition ID #\n20WD40666\nJob Title\nSe...     4.0   
601  Secure our Nation, Ignite your Future\n\nJob S...     4.2   
640  About Us\n\nAt GutCheck, we pioneered agile ma...     3.8   
641  Sr. ML/Data Scientist - AI/NLP/Chatbot\n\nMist...     3.8   

           Company Name           Location      Headquarters  \
0           Healthfirst       New York, NY      New York, NY   
32     Acuity Insurance      Sheboygan, WI     Sheboygan, WI   
38          Edmunds.com   Santa Monica, CA  Santa Monica, CA   
45              Klaviyo         Boston, MA        Boston, MA   
54          Benson Hill    Saint Louis, MO   Saint Louis, MO   
..                  ...                ...               ...   
579  Affinity Solutions       New York, NY      New York, NY   
583            Autodesk  San Francisco, CA    San Rafael, CA   
601             ManTech     Alexandria, VA       Herndon, VA   
640            GutCheck         Denver, CO        Denver, CO   
641    Juniper Networks      Cupertino, CA     Sunnyvale, CA   

                        Size       Type of ownership  \
0     1001 to 5000 employees  Nonprofit Organization   
32    1001 to 5000 employees       Company - Private   
38     501 to 1000 employees       Company - Private   
45      201 to 500 employees       Company - Private   
54      201 to 500 employees       Company - Private   
..                       ...                     ...   
579      51 to 200 employees       Company - Private   
583  5001 to 10000 employees        Company - Public   
601  5001 to 10000 employees        Company - Public   
640      51 to 200 employees       Company - Private   
641  5001 to 10000 employees        Company - Public   

                         Industry  ... company_age python  excel  hadoop  \
0              Insurance Carriers  ...        27.0      0      0       0   
32             Insurance Carriers  ...        95.0      0      0       0   
38                       Internet  ...        54.0      1      1       0   
45   Computer Hardware & Software  ...         8.0      0      0       0   
54      Biotech & Pharmaceuticals  ...         8.0      1      1       0   
..                            ...  ...         ...    ...    ...     ...   
579       Advertising & Marketing  ...        22.0      1      0       1   
583  Computer Hardware & Software  ...        38.0      1      0       1   
601        Research & Development  ...        52.0      1      0       1   
640       Advertising & Marketing  ...        11.0      0      0       0   
641   Telecommunications Services  ...        24.0      1      1       0   

     spark aws  tableau  big_data        job_simp  seniority  


d. A partir de la columna “Salary Estimate”, genera dos columnas: Salario Estimado Mínimo
y Máximo. (hint: Utiliza el método apply sobre la columna.)


In [132]:
df_jobs[['Salario Estimado Mínimo', 'Salario Estimado Máximo']] = df_jobs['Salary Estimate'].apply(
    lambda salary_string: pd.Series([int(s.replace('$', '').replace('K', '').strip()) for s in salary_string.split('-')])
    if '-' in salary_string
    else pd.Series([int(s.replace('$', '').replace('K', '').strip()), int(s.replace('$', '').replace('K', '').strip())])
)
df_jobs

Job Title Salary Estimate  \
0                               Sr Data Scientist        137-171    
32   Senior Research Statistician- Data Scientist         75-131    
38                  Senior Analyst/Data Scientist         75-131    
45                          Senior Data Scientist         75-131    
54                          Senior Data Scientist         75-131    
..                                            ...             ...   
579                          Senior Data Engineer        138-158    
583                         Senior Data Scientist         80-132    
601                         Senior Data Scientist         87-141    
640                         Senior Data Scientist        105-167    
641        Sr. ML/Data Scientist - AI/NLP/Chatbot        105-167    

                                       Job Description  Rating  \
0    Description\n\nThe Senior Data Scientist is re...     3.1   
32   Acuity is seeking a Senior Research Statistici...     4.8   
38   At Edmunds were driven to make car buying easi...     3.4   
45   Klaviyo is looking for Senior Data Scientists ...     4.8   
54   Benson Hill empowers innovators to develop mor...     3.5   
..                                                 ...     ...   
579  Senior Data Engineer\n\nMaster’s degree in Inf...     2.9   
583  Job Requisition ID #\n20WD40666\nJob Title\nSe...     4.0   
601  Secure our Nation, Ignite your Future\n\nJob S...     4.2   
640  About Us\n\nAt GutCheck, we pioneered agile ma...     3.8   
641  Sr. ML/Data Scientist - AI/NLP/Chatbot\n\nMist...     3.8   

           Company Name           Location      Headquarters  \
0           Healthfirst       New York, NY      New York, NY   
32     Acuity Insurance      Sheboygan, WI     Sheboygan, WI   
38          Edmunds.com   Santa Monica, CA  Santa Monica, CA   
45              Klaviyo         Boston, MA        Boston, MA   
54          Benson Hill    Saint Louis, MO   Saint Louis, MO   
..                  ...                ...               ...   
579  Affinity Solutions       New York, NY      New York, NY   
583            Autodesk  San Francisco, CA    San Rafael, CA   
601             ManTech     Alexandria, VA       Herndon, VA   
640            GutCheck         Denver, CO        Denver, CO   
641    Juniper Networks      Cupertino, CA     Sunnyvale, CA   

                        Size       Type of ownership  \
0     1001 to 5000 employees  Nonprofit Organization   
32    1001 to 5000 employees       Company - Private   
38     501 to 1000 employees       Company - Private   
45      201 to 500 employees       Company - Private   
54      201 to 500 employees       Company - Private   
..                       ...                     ...   
579      51 to 200 employees       Company - Private   
583  5001 to 10000 employees        Company - Public   
601  5001 to 10000 employees        Company - Public   
640      51 to 200 employees       Company - Private   
641  5001 to 10000 employees        Company - Public   

                         Industry  ... excel hadoop  spark  aws  tableau  \
0              Insurance Carriers  ...     0      0      0    1        0   
32             Insurance Carriers  ...     0      0      0    0        0   
38                       Internet  ...     1      0      0    1        1   
45   Computer Hardware & Software  ...     0      0      0    0        0   
54      Biotech & Pharmaceuticals  ...     1      0      0    0        0   
..                            ...  ...   ...    ...    ...  ...      ...   
579       Advertising & Marketing  ...     0      1      1    0        1   
583  Computer Hardware & Software  ...     0      1      1    0        0   
601        Research & Development  ...     0      1      1    0        1   
640       Advertising & Marketing  ...     0      0      0    0        0   
641   Telecommunications Services  ...     1      0      1    1        0   

    big_data        job_simp  seniority  Salario Estimado Mínim

e. Realiza la recodificación de la columna Size con los valores de la siguiente tabla: (hint:
utilice reemplazo con diccionario usando el método replace sobre la columna.)
Size                          Descripción
10000+ employees          Mega Empresas
5001 to 10000 employees   Grandes Empresas
1001 to 5000 employees    Medianas Empresas
201 to 500 employees      Pequeñas Empresas
51 to 200 employees       Pequeñas Grandes Empresas
501 to 1000 employees     Microempresas
Unknown Empresas sin Información

In [138]:
#Recodificación de la columna 'Size'
size_mapping = {
    '10000+ employees': 'Mega Empresas',
    '5001 to 10000 employees': 'Grandes Empresas',
    '1001 to 5000 employees': 'Medianas Empresas',
    '201 to 500 employees': 'Pequeñas Empresas',
    '51 to 200 employees': 'Pequeñas Grandes Empresas',
    '501 to 1000 employees': 'Microempresas',
    'Unknown': 'Empresas sin Información'
}
df_jobs['Size'] = df_jobs['Size'].replace(size_mapping)
df_jobs

Job Title Salary Estimate  \
0                               Sr Data Scientist        137-171    
32   Senior Research Statistician- Data Scientist         75-131    
38                  Senior Analyst/Data Scientist         75-131    
45                          Senior Data Scientist         75-131    
54                          Senior Data Scientist         75-131    
..                                            ...             ...   
579                          Senior Data Engineer        138-158    
583                         Senior Data Scientist         80-132    
601                         Senior Data Scientist         87-141    
640                         Senior Data Scientist        105-167    
641        Sr. ML/Data Scientist - AI/NLP/Chatbot        105-167    

                                       Job Description  Rating  \
0    Description\n\nThe Senior Data Scientist is re...     3.1   
32   Acuity is seeking a Senior Research Statistici...     4.8   
38   At Edmunds were driven to make car buying easi...     3.4   
45   Klaviyo is looking for Senior Data Scientists ...     4.8   
54   Benson Hill empowers innovators to develop mor...     3.5   
..                                                 ...     ...   
579  Senior Data Engineer\n\nMaster’s degree in Inf...     2.9   
583  Job Requisition ID #\n20WD40666\nJob Title\nSe...     4.0   
601  Secure our Nation, Ignite your Future\n\nJob S...     4.2   
640  About Us\n\nAt GutCheck, we pioneered agile ma...     3.8   
641  Sr. ML/Data Scientist - AI/NLP/Chatbot\n\nMist...     3.8   

           Company Name           Location      Headquarters  \
0           Healthfirst       New York, NY      New York, NY   
32     Acuity Insurance      Sheboygan, WI     Sheboygan, WI   
38          Edmunds.com   Santa Monica, CA  Santa Monica, CA   
45              Klaviyo         Boston, MA        Boston, MA   
54          Benson Hill    Saint Louis, MO   Saint Louis, MO   
..                  ...                ...               ...   
579  Affinity Solutions       New York, NY      New York, NY   
583            Autodesk  San Francisco, CA    San Rafael, CA   
601             ManTech     Alexandria, VA       Herndon, VA   
640            GutCheck         Denver, CO        Denver, CO   
641    Juniper Networks      Cupertino, CA     Sunnyvale, CA   

                          Size       Type of ownership  \
0            Medianas Empresas  Nonprofit Organization   
32           Medianas Empresas       Company - Private   
38               Microempresas       Company - Private   
45           Pequeñas Empresas       Company - Private   
54           Pequeñas Empresas       Company - Private   
..                         ...                     ...   
579  Pequeñas Grandes Empresas       Company - Private   
583           Grandes Empresas        Company - Public   
601           Grandes Empresas        Company - Public   
640  Pequeñas Grandes Empresas       Company - Private   
641           Grandes Empresas        Company - Public   

                         Industry  ... excel hadoop  spark  aws  tableau  \
0              Insurance Carriers  ...     0      0      0    1        0   
32             Insurance Carriers  ...     0      0      0    0        0   
38                       Internet  ...     1      0      0    1        1   
45   Computer Hardware & Software  ...     0      0      0    0        0   
54      Biotech & Pharmaceuticals  ...     1      0      0    0        0   
..                            ...  ...   ...    ...    ...  ...      ...   
579       Advertising & Marketing  ...     0      1      1    0        1   
583  Computer Hardware & Software  ...     0      1      1    0        0   
601        Research & Development  ...     0      1      1    0        1   
640       Advertising & Marketing  ...     0      0      0    0        0   
641   Telecommunications Services  ...     1      0      1    1        0   

    big_data        job_simp  seniority

f. Finalmente, genera una tabla pivote que muestre la media del salario estimado mínimo y
la media del salario estimado máximo por tamaño de empresa. (hint: utiliza
pd.pivot_table para generar la vista adecuada con las columnas generadas.)

In [140]:

# Generar la tabla pivote
pivot_table_size_salary = pd.pivot_table(df_jobs, index='Size',
                                         values=['Salario Estimado Mínimo', 'Salario Estimado Máximo'],
                                         aggfunc='mean')
pivot_table_size_salary

Salario Estimado Máximo  Salario Estimado Mínimo
Size                                                                       
Empresas sin Información                110.500000                73.000000
Grandes Empresas                        138.875000                92.125000
Medianas Empresas                       137.461538                93.923077
Mega Empresas                           151.111111                97.888889
Microempresas                           146.235294               100.176471
Pequeñas Empresas                       141.142857                93.571429
Pequeñas Grandes Empresas               137.666667               100.666667